pkg/client/cache/listers.go

```diff
@@ -77,20 +77,33 @@ type storePodsNamespacer struct {
 // that.
 func (s storePodsNamespacer) List(selector labels.Selector) (pods api.PodList, err error) {
 	list := api.PodList{}
-	for _, m := range s.store.List() {
-		pod := m.(*api.Pod)
-		if s.namespace == api.NamespaceAll || s.namespace == pod.Namespace {
+	if s.namespace == api.NamespaceAll {
+		for _, m := range s.indexer.List() {
+			pod := m.(*api.Pod)
 			if selector.Matches(labels.Set(pod.Labels)) {
 				list.Items = append(list.Items, *pod)
 			}
 		}
+		return list, nil
+	}
+
+	key := &api.Pod{ObjectMeta: api.ObjectMeta{Namespace: s.namespace}}
+	items, err := s.indexer.Index(NamespaceIndex, key)
```

Comment:
If this index doesn't exist, fall back to the old behavior and print a `glog.Warning` so we can find the offenders.